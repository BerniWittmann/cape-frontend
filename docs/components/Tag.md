# Tag

A Component to view a single Tag

## Props

<!-- @vuese:Tag:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|tag|The Tag object|`Object`|`true`|-|
|closable|Option whether the tag should be deleteable or not|`Boolean`|`false`|false|
|size|Size of the Tag|`'medium'` / `'small'` / `'mini'`|`false`|-|

<!-- @vuese:Tag:props:end -->


## Events

<!-- @vuese:Tag:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|close|Emitted when the close button is clicked|-|

<!-- @vuese:Tag:events:end -->


