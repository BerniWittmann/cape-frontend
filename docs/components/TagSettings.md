# TagSettings

A Settings Tab to edit the tags

## Methods

<!-- @vuese:TagSettings:methods:start -->
|Method|Description|Parameters|
|---|---|---|
|updateLayoutTable|Update the Table Layout|-|

<!-- @vuese:TagSettings:methods:end -->


