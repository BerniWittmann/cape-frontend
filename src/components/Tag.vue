<template>
  <el-tag class="tag" type="info" :style="style" :closable="closable" :size="size" @close="emitClose">{{ tag.name }}</el-tag>
</template>

<script>
import { convertHexToRgba } from '@/utils/helpers'

/*
 * @vuese
 * @group Components
 *
 * A Component to view a single Tag
 */
export default {
  name: 'Tag',
  props: {
    // The Tag object
    tag: {
      type: Object,
      required: true
    },

    // Option whether the tag should be deleteable or not
    closable: {
      type: Boolean,
      default: false
    },

    // Size of the Tag
    size: {
      // `'medium'` / `'small'` / `'mini'`
      type: String
    }
  },

  computed: {
    style() {
      if (!this.tag.color) return {}

      return {
        backgroundColor: convertHexToRgba(this.tag.color, 0.1),
        borderColor: convertHexToRgba(this.tag.color, 0.2),
        color: convertHexToRgba(this.tag.color, 1)
      }
    }
  },

  methods: {
    emitClose() {
      // Emitted when the close button is clicked
      this.$emit('close')
    }
  }
}

</script>

<style lang="scss">
.tag {
  margin-right: 5px;
  &:last-of-type {
    margin-right: 0;
  }

  .el-tag__close {
    color: inherit !important;
    &:hover {
      background-color: inherit !important;
      color: #fff;
    }
  }
}
</style>
