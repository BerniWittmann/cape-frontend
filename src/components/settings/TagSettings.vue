<template>
  <el-row type="flex" justify="center">
    <el-col :span="8">
      <el-row>
        <el-col>
          <tag class="tag-row__tag" :tag="tag" v-for="tag in tags" :key="tag.id" closable @close="deleteTag(tag)"></tag>
        </el-col>
      </el-row>
      <br>

      <el-row>
        <el-col>
          <tag :tag="tag" v-if="tag.name && tag.name.length > 0"></tag>
        </el-col>
      </el-row>

      <el-form ref="tagForm" :model="tag" label-position="top" :rules="rules">
        <el-form-item :label="$t('settings.tag.name')" prop="name">
          <el-input v-model="tag.name"></el-input>
        </el-form-item>
        <el-form-item :label="$t('settings.tag.color')" prop="color">
          <el-color-picker v-model="tag.color"></el-color-picker>
        </el-form-item>
        <el-form-item>
          <el-button type="success" @click.native="submit">{{ $t('settings.tag.submit') }}</el-button>
          <el-button type="danger" plain @click.native="reset">{{ $t('settings.tag.reset') }}</el-button>
        </el-form-item>
      </el-form>
    </el-col>
  </el-row>
</template>

<script>
/* ============
 * Tag Settings Component
 * ============
 *
 * A component to edit the Tags
 */

import { DEFAULT_TAG_COLOR } from '@/utils/constants'
import Tag from '@/models/tag'
import tagService from '@/services/tag'

import TagComponent from '@/components/Tag.vue'

export default {
  components: {
    Tag: TagComponent
  },

  data() {
    return {
      tag: {
        name: undefined,
        color: DEFAULT_TAG_COLOR
      },
      rules: {
        name: [
          { required: true, message: this.$t('settings.tag.validation.name.required'), trigger: 'blur' }
        ],
        color: [
          { required: true, message: this.$t('settings.tag.validation.color.required'), trigger: 'blur' }
        ]
      }
    }
  },

  methods: {
    submit() {
      this.$refs.tagForm.validate((valid) => {
        if (valid) {
          const tag = new Tag(this.tag)
          tagService.create(tag).then(() => {
            this.reset()
          })
        }
      })
    },

    reset() {
      this.tag = {
        name: undefined,
        color: DEFAULT_TAG_COLOR
      }
      this.$refs.tagForm.resetFields()
    },

    deleteTag(tag) {
      tagService.remove(tag)
    }
  },

  computed: {
    tags() {
      return this.$store.state.tag.tags
    }
  }
}

</script>

<style scoped lang="scss">
.tag-row__tag {
  margin-bottom: 5px;
}
</style>
