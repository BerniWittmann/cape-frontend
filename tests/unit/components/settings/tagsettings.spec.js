import { mount } from '@vue/test-utils'
import { i18n } from '../../setupPlugins'
import { ColorPicker } from 'element-ui'

import TagSettings from '@/components/settings/TagSettings.vue'
import Tag from '@/components/Tag.vue'

import tagService from '@/services/tag'

describe('Components', () => {
  describe('TagSettings', () => {
    let store
    let cmp

    tagService.remove = jest.fn()
    tagService.create = jest.fn().mockImplementation(() => ({
      then: (arg) => arg()
    }))

    beforeEach((done) => {
      store = {
        state: {
          tag: {
            tags: [{
              id: '142',
              name: 'My Tag',
              color: '#FF0000'
            }, {
              id: '3',
              name: 'My Second Tag',
              color: '#FF00FF'
            }]
          }
        }
      }
      render(done)
    })

    function render(done) {
      cmp = mount(TagSettings, {
        i18n,
        mocks: {
          $store: store
        },
        sync: false
      })
      cmp.vm.$nextTick(done)
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    it('renders the available Tags', () => {
      const tags = cmp.findAll(Tag)
      expect(tags.length).toEqual(2)
      expect(tags.at(0).props('tag')).toEqual(store.state.tag.tags[0])
      expect(tags.at(1).props('tag')).toEqual(store.state.tag.tags[1])
    })
    it('the existing tags can be deleted', () => {
      const tags = cmp.findAll(Tag)
      expect(tags.length).toEqual(2)
      expect(tags.at(0).props('closable')).toBeTruthy()
      expect(tags.at(1).props('tag')).toBeTruthy()

      tags.at(1).vm.$emit('close')

      expect(tagService.remove).toHaveBeenCalledWith(store.state.tag.tags[1])
    })
    it('does not display the tag preview if no name given', () => {
      const tags = cmp.findAll(Tag)
      expect(tags.length).toEqual(2)
    })
    it('displays a preview of the tag', (done) => {
      cmp.find('input').setValue('My Name')
      cmp.vm.tag.color = '#FFFF00'

      cmp.vm.$nextTick(() => {
        const tags = cmp.findAll(Tag)
        expect(tags.length).toEqual(3)
        expect(tags.at(2).props('tag')).toEqual({
          color: '#FFFF00',
          name: 'My Name'
        })
        done()
      })
    })
    it('shows an error if no name given on submit', (done) => {
      console.warn = jest.fn()
      cmp.vm.submit()

      cmp.vm.$nextTick(() => {
        const warning = cmp.find('.el-form-item__error')
        expect(warning.exists()).toBeTruthy()
        expect(warning.text()).toContain('settings.tag.validation.name.required')
        done()
      })
    })
    it('shows an error if no color given on submit', (done) => {
      console.warn = jest.fn()
      cmp.find('input').setValue('My Name')
      cmp.vm.tag.color = undefined
      cmp.vm.submit()

      cmp.vm.$nextTick(() => {
        const warning = cmp.find('.el-form-item__error')
        expect(warning.exists()).toBeTruthy()
        expect(warning.text()).toContain('settings.tag.validation.color.required')
        done()
      })
    })
    it('shows a color picker', () => {
      expect(cmp.contains(ColorPicker)).toBeTruthy()
    })
    it('validates the form on submit', (done) => {
      cmp.vm.submit()

      cmp.vm.$nextTick(() => {
        expect(cmp.contains('.el-form-item__error')).toBeTruthy()
        done()
      })
    })
    it('the form has a submit and a cancel button', () => {
      expect(cmp.findAll('.el-form-item > .el-form-item__content > .el-button').length).toEqual(2)
    })
    it('the form can be reset', () => {
      cmp.find('input').setValue('My Name')
      cmp.vm.tag.color = '#FFF000'
      expect(cmp.vm.tag.name).toEqual('My Name')

      cmp.vm.reset()

      expect(cmp.vm.tag.name).toEqual(undefined)
      expect(cmp.vm.tag.color).toEqual('#6CC9E8')
    })
    it('the form can be submitted', (done) => {
      cmp.find('input').setValue('My new Name')
      cmp.vm.tag.color = '#FFF000'
      cmp.vm.submit()

      cmp.vm.$nextTick(() => {
        expect(tagService.create).toHaveBeenCalledWith({
          _id: undefined,
          color: '#FFF000',
          name: 'My new Name'
        })
        done()
      })
    })
    it('the form is reset after submitting', (done) => {
      cmp.find('input').setValue('My new Name')
      cmp.vm.tag.color = '#FFF000'
      cmp.vm.submit()

      cmp.vm.$nextTick(() => {
        expect(cmp.vm.tag.color).toEqual('#6CC9E8')
        expect(cmp.vm.tag.name).toEqual(undefined)
        done()
      })
    })
  })
})
