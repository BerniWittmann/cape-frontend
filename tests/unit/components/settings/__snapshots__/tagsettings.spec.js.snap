// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components TagSettings renders 1`] = `
<section class="el-container">
  <aside class="el-aside" style="width: 180px;">
    <div class="el-table el-table---disable-row-hover el-table--fit el-table--scrollable-x el-table--enable-row-hover el-table--enable-row-transition" style="height: 420px;">
      <div class="hidden-columns">
        <div></div>
      </div>
      <div class="el-table__header-wrapper">
        <table cellspacing="0" cellpadding="0" border="0" class="el-table__header" style="width: 80px;">
          <colgroup>
            <col name="el-table_1_column_1" width="80">
          </colgroup>
          <thead class="">
            <tr class="">
              <th colspan="1" rowspan="1" class="el-table_1_column_1     is-leaf" style="border: 0px; padding: 0px;">
                <div class="cell">
                  <div class="el-input el-input--mini">
                    <!----><input type="text" autocomplete="off" placeholder="settings.tag.search" class="el-input__inner">
                    <!---->
                    <!---->
                    <!---->
                  </div>
                </div>
              </th>
            </tr>
          </thead>
        </table>
      </div>
      <div class="el-table__body-wrapper is-scrolling-left">
        <table cellspacing="0" cellpadding="0" border="0" class="el-table__body" style="width: 80px;">
          <colgroup>
            <col name="el-table_1_column_1" width="80">
          </colgroup>
          <tbody>
            <tr class="el-table__row">
              <td rowspan="1" colspan="1" class="el-table_1_column_1  " style="border: 0px; padding-right: 5px; padding-top: 0px; padding-bottom: 5px; float: right;">
                <div class="cell"><span class="tag el-tag el-tag--info" style="background-color: rgba(255, 0, 0, 0.1); color: rgb(255, 0, 0);" name="el-zoom-in-center">My Tag</span></div>
              </td>
            </tr>
            <tr class="el-table__row">
              <td rowspan="1" colspan="1" class="el-table_1_column_1  " style="border: 0px; padding-right: 5px; padding-top: 0px; padding-bottom: 5px; float: right;">
                <div class="cell"><span class="tag el-tag el-tag--info" style="background-color: rgba(255, 0, 255, 0.1); color: rgb(255, 0, 255);" name="el-zoom-in-center">My Second Tag</span></div>
              </td>
            </tr>
            <!---->
          </tbody>
        </table>
        <!---->
        <!---->
      </div>
      <!---->
      <!---->
      <!---->
      <!---->
      <div class="el-table__column-resize-proxy" style="display: none;"></div>
    </div>
    <div role="switch" class="el-switch"><input type="checkbox" name="" true-value="true" class="el-switch__input">
      <!----><span class="el-switch__core" style="width: 40px;"></span><span class="el-switch__label el-switch__label--right"><!----><span aria-hidden="true">settings.tag.deletable</span></span></div>
  </aside>
  <main class="el-main">
    <h3>settings.tag.new</h3> <span class="tag el-tag el-tag--info" style="background-color: rgba(108, 201, 232, 0.1); color: rgb(108, 201, 232);" name="el-zoom-in-center"></span>
    <form class="el-form el-form--label-top">
      <div class="el-form-item is-required"><label for="name" class="el-form-item__label">settings.tag.name</label>
        <div class="el-form-item__content">
          <div class="el-input">
            <!----><input type="text" autocomplete="off" class="el-input__inner">
            <!---->
            <!---->
            <!---->
          </div>
          <!---->
        </div>
      </div>
      <div class="el-form-item is-required"><label for="color" class="el-form-item__label">settings.tag.color</label>
        <div class="el-form-item__content">
          <div class="el-color-picker">
            <!---->
            <div class="el-color-picker__trigger"><span class="el-color-picker__color"><span class="el-color-picker__color-inner" style="background-color: rgb(108, 201, 232);"></span>
              <!----></span><span class="el-color-picker__icon el-icon-arrow-down"></span></div>
            <div class="el-color-dropdown el-color-picker__panel" style="display: none;" name="el-zoom-in-top">
              <div class="el-color-dropdown__main-wrapper">
                <div class="el-color-hue-slider is-vertical" style="float: right;">
                  <div class="el-color-hue-slider__bar"></div>
                  <div class="el-color-hue-slider__thumb" style="left: 0px; top: 0px;"></div>
                </div>
                <div class="el-color-svpanel">
                  <div class="el-color-svpanel__white"></div>
                  <div class="el-color-svpanel__black"></div>
                  <div class="el-color-svpanel__cursor" style="top: 0px; left: 0px;">
                    <div></div>
                  </div>
                </div>
              </div>
              <!---->
              <!---->
              <div class="el-color-dropdown__btns"><span class="el-color-dropdown__value"><div class="el-input el-input--mini"><!----><input type="text" autocomplete="off" class="el-input__inner"><!----><!----><!----></div></span><button type="button" class="el-button el-color-dropdown__link-btn el-button--text el-button--mini">
                  <!---->
                  <!----><span>
        清空
      </span></button><button type="button" class="el-button el-color-dropdown__btn el-button--default el-button--mini is-plain">
                  <!---->
                  <!----><span>
        确定
      </span></button></div>
            </div>
          </div>
          <!---->
        </div>
      </div>
      <div class="el-form-item">
        <!---->
        <div class="el-form-item__content"><button type="button" class="el-button el-button--success">
            <!---->
            <!----><span>settings.tag.submit</span></button> <button type="button" class="el-button el-button--danger is-plain">
            <!---->
            <!----><span>settings.tag.reset</span></button>
          <!---->
        </div>
      </div>
    </form>
  </main>
</section>
`;
